velox/expression/LambdaExpr.cpp

```diff
@@ -129,10 +139,46 @@ class ExprCallable : public Callable {
   RowTypePtr signature_;
   RowVectorPtr capture_;
   std::shared_ptr<Expr> body_;
+  // List of Shared Exprs that are decendants of 'body_' for which reset() needs
+  // to be called before calling `body_->eval()`.
+  std::vector<std::shared_ptr<Expr>> sharedExprsToReset_;
 };
 
+void extractSharedExpressions(
+    const ExprPtr& expr,
+    std::unordered_set<ExprPtr>& shared) {
+  for (const auto& input : expr->inputs()) {
+    if (input->isMultiplyReferenced()) {
+      shared.insert(input);
```

Comment:
perhaps, we don't need to recurse here and can just 'continue'

Reply:
all children of a shared sub expression would also be shared and we should reset them too. If we dont then the parent would be reset, but the children would still have retained shared results and can cause the same issue in the next invocation.